### PR TITLE
Added Discord RPC Module

### PR DESCRIPTION
**Pull request by humboldt123**

### CODE REVIEW

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    @Setting(name = "Server")
+    private var server = true
+
+    @Setting(name = "Flair")
+    private var message = true
+
+    var lastUpdate: Long = 0L
+    var rpcEnabled = false
+
+    val messages = arrayOf(
```
Comment (dominikaaaa) [inline src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:33]: these should be user configurable

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    )
+
+    override fun onEnable() {
+        rpcEnabled = true
+        initDiscord()
+    }
+
+    override fun onDisable() {
+        DiscordRPC.discordShutdown()
+        rpcEnabled = false
```
Comment (dominikaaaa) [inline src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:57]: why is this needed?

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        println("downdoot")
```
Comment (dominikaaaa) [inline src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:58]: and this?

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    private var username = true
+
+    @Setting(name = "Server")
+    private var server = true
+
+    @Setting(name = "Flair")
+    private var message = true
+
+    var lastUpdate: Long = 0L
+    var rpcEnabled = false
```
Comment (dominikaaaa) [inline src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:31]: why is this needed?

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+    @EventHandler
+    private val updateListener =
+            Listener<TickEvent.Client.InGame>({
+                if (lastUpdate == 0L) lastUpdate = System.currentTimeMillis()
+                if ((lastUpdate + 15000L <= System.currentTimeMillis()) && rpcEnabled) { // 15 seconds, should be good.
+                    DiscordRPC.discordRunCallbacks()
+                    var bottomString = ""
+                    var topString = ""
+                    if (message && server) bottomString = ("127.0.0.1 (ez doxx)") + " | " + messages.random()
```
Comment (dominikaaaa) [inline src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:69]: should be user configurable

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+                    else if (message) bottomString = messages.random()
+                    else if (server) bottomString = ("127.0.0.1 (ez doxx)")
+
+                    if (version && username) topString = (KamiMod.MODVER) + " | Notch"
```
Comment (dominikaaaa) [inline src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:73]: why????

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+                    else if (version) topString = messages.random()
+                    else if (username) topString = ("Notch")
```
Comment (dominikaaaa) [inline src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:75]: what??

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+                    presence.setDetails(topString) // This is the top half with modver and username
+                    presence.setBigImage("kami", "kamiclient.com") // Image key (kami) and text when you scroll over it (kamiclient.com)
+                    DiscordRPC.discordUpdatePresence(presence.build())
+                    lastUpdate += 15000L
+                }
+
+
+            })
+
+    // Yes, its the same code but again. Yes, I know its bad practice but at the moment I don't care.
```
Comment (dominikaaaa) [inline src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:88]: fix this

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        val handlers = DiscordEventHandlers.Builder().setReadyEventHandler { user: DiscordUser ->
+            val presence = DiscordRichPresence.Builder(bottomString)
+            presence.setDetails(topString)
+            DiscordRPC.discordUpdatePresence(presence.build())
+        }.build()
+
+        DiscordRPC.discordInitialize("753664640789118999", handlers, false);
+        DiscordRPC.discordRegister("753664640789118999", "");
+        DiscordRPC.discordRunCallbacks()
+        println("lol")
```
Comment (dominikaaaa) [inline src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:109]: why?

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    fun getOperatingSystemMessage() = when (getOperatingSystem()) {
```
Comment (dominikaaaa) [inline src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:112]: why is this here????

**Review by dominikaaaa — CHANGES_REQUESTED** [review]

**Comment by dominikaaaa** [comment]
see VarMap for variables to use

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt (hunk 1 above)
Comment (humboldt123) [reply src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:33]: No, the reason its what it is is discord's reccomended limit to stop api spam.

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt (hunk 3 above)
Comment (humboldt123) [reply src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:58]: That was a debug line I forgot to remove, I wanted 086's feedback before it becomes final.

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt (hunk 4 above)
Comment (humboldt123) [reply src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:31]: It's needed to prevent API Spam

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt (hunk 5 above)
Comment (humboldt123) [reply src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:69]: Yes, 086 said he would add them himself.

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt (hunk 6 above)
Comment (humboldt123) [reply src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:73]: Yes, 086 said he would add them himself.

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt (hunk 7 above)
Comment (humboldt123) [reply src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:75]: Yes, 086 said he would add them himself.

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt (hunk 8 above)
Comment (humboldt123) [reply src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:88]: Yeah, can do.

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt (hunk 9 above)
Comment (humboldt123) [reply src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:109]: Debug 👏 Line

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt (hunk 10 above)
Comment (humboldt123) [reply src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:112]: For one of the messages

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt (hunk 1 above)
Comment (dominikaaaa) [reply src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:33]: what?????

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt (hunk 4 above)
Comment (dominikaaaa) [reply src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:31]: how?

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt (hunk 10 above)
Comment (dominikaaaa) [reply src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:112]: use the utils

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        else if (username) topString = ("Notch")
+
+        val handlers = DiscordEventHandlers.Builder().setReadyEventHandler { user: DiscordUser ->
+            val presence = DiscordRichPresence.Builder(bottomString)
+            presence.setDetails(topString)
+            DiscordRPC.discordUpdatePresence(presence.build())
+        }.build()
+
+        DiscordRPC.discordInitialize("753664640789118999", handlers, false);
+        DiscordRPC.discordRegister("753664640789118999", "");
```
Comment (zeroeightysix) [inline src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:107]: What's this magic number? I'm assuming it's some kind of ID for the rich presence 'application', or whatever discord calls them - can you create a `private const val` for it?

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+        DiscordRPC.discordRunCallbacks()
```
Comment (zeroeightysix) [inline src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:108]: What callbacks have we registered? Unless we're using callbacks, I don't think there's a reason to call this method.

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt (hunk 10 above)
Comment (zeroeightysix) [reply src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:112]: What utils? If you're talking about the `VarMap`, I'll do that myself. Integrating it into the UI will be quite tricky, so I can't quite yet hope for @humboldt123 to be able to do that himself.

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+        DiscordRPC.discordShutdown()
+        rpcEnabled = false
+        println("downdoot")
+    }
+
+    @EventHandler
+    private val updateListener =
+            Listener<TickEvent.Client.InGame>({
+                if (lastUpdate == 0L) lastUpdate = System.currentTimeMillis()
+                if ((lastUpdate + 15000L <= System.currentTimeMillis()) && rpcEnabled) { // 15 seconds, should be good.
```
Comment (zeroeightysix) [inline src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:65]: `rpcEnabled` will always be `true`. Modules are dynamically (un)subscribed from the event bus when they are enabled or disabled - there's no need to keep around the enabled state of the module.

Just remove the variable entirely.

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+                    DiscordRPC.discordRunCallbacks()
+                    var bottomString = ""
+                    var topString = ""
+                    if (message && server) bottomString = ("127.0.0.1 (ez doxx)") + " | " + messages.random()
+                    else if (message) bottomString = messages.random()
+                    else if (server) bottomString = ("127.0.0.1 (ez doxx)")
```
Comment (zeroeightysix) [inline src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:71]: Declaration and assignment of `bottomString` can be merged here. Use a nifty `if` expression.

Same goes for `topString`.

**Review by zeroeightysix — CHANGES_REQUESTED** [review]
Obviously also fix the duplicate code that @dominikaaaa pointed out. Once workable, I'll move the settings to those similar to the HUD.

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt (hunk 11 above)
Comment (dominikaaaa) [reply src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:107]: Should be user customizable imo

Comment (humboldt123) [reply src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:107]: Yes, it's the application id. I'll create a variable for it.

Comment (dominikaaaa) [reply src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:107]: If you do that, you need to make the names for the image icons customizable as well, since those correlate with the uploaded names on your discord application (not the discord client, the one you create for your rpc that gives you the ID)

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+                    val presence = DiscordRichPresence.Builder("${server to "127.0.0.1 (ez doxx)"}${(server && message) to " | "}${message to messages.random()}")
+                            // This is the top half with modver and username
+                            .setDetails("${username to "Notch"}${(username && version) to " | "}${version to KamiMod.MODVER}")
+                            // Image key (kami) and text when you scroll over it (kamiclient.com)
+                            .setBigImage("kami", "kamiclient.com")
+                            // Small image key and text when hovered over
+                            .setSmallImage("bigrat", "he is massive :)")
+
+                    //Build it
+                    DiscordRPC.discordUpdatePresence(presence.build())
```
Comment (zeroeightysix) [inline src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:78]: The significant action here is definitely updating the RPC, not building the information

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    override fun onDisable() {
+        DiscordRPC.discordShutdown()
+    }
+
+    @EventHandler
+    private val updateListener =
+            Listener<TickEvent.Client.InGame>({
+                if (lastUpdate == 0L) lastUpdate = System.currentTimeMillis()
```
Comment (zeroeightysix) [inline src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:63]: If the user disables DiscordRPC for, let's say, 30 minutes and then re-enables it, this check won't actually work (as `lastUpdate` was set to the current time 30 minutes ago).

`lastUpdate` will be really behind then, and will actually cause this event listener to 'catch' up by spamming a RPC update to discord each of the following **120** ticks!! (30 mins = 1800 seconds = 120 x 15 seconds)

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+                    lastUpdate += updateLimit
+                }
+            })
+
+    fun initDiscord() {
+        val handlers = DiscordEventHandlers.Builder().setReadyEventHandler { user: DiscordUser ->
+            println("${user.username} is using the RPC.")
+            val presence = DiscordRichPresence.Builder("Loading...")
+            DiscordRPC.discordUpdatePresence(presence.build())
+        }.build()
```
Comment (zeroeightysix) [inline src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:88]: To be honest, this isn't really needed. I'd rather have the next tick update the RPC instead of next 15 seconds being "loading" the RPC.

**Review by zeroeightysix — CHANGES_REQUESTED** [review]

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    override fun onDisable() {
+        DiscordRPC.discordShutdown()
+    }
+
+    @EventHandler
+    private val updateListener =
+            Listener<TickEvent.Client.InGame>({
+                if (lastUpdate == 0L) lastUpdate = System.currentTimeMillis()
```
Comment (zeroeightysix) [inline src/main/kotlin/me/zeroeightsix/kami/feature/module/DiscordStatus.kt:62]: `lastUpdate` will never be `0` by the time the tick listener is called, making this dead code.

**Review by zeroeightysix — CHANGES_REQUESTED** [review]

**Review by zeroeightysix — APPROVED** [review]
